File: projects/java/src/main/java/com/stater/statemachine/java/StaterStateMachine.java
package com.stater.statemachine.java;

import java.io.Console;
import java.text.Collator;
import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;


public abstract class StaterStateMachine<T, C extends Context> {
    private C context;
    private final T startState;
    private final Set<T> states;
    private T state;
    private final List<Transition<T, C>> transitions;
    private Map<T, List<Transition<T, C>>> transitionsGroupedStart = new HashMap<>();
    private Map<String, Transition<T, C>> transitionsByName = new HashMap<>();
    private Map<String, List<TransitionMiddleware<C>>> transitionMiddlewares = new HashMap<>();
    private List<TransitionNameMiddleware<C>> transitionAllMiddlewares = new ArrayList<>();
    private Map<String, List<Event<C>>> transitionCallbacks = new HashMap<>();
    private List<NameEvent<C>> transitionAllCallbacks = new ArrayList<>();
    private Map<T, List<Event<C>>> stateCallbacks = new HashMap<>();
    private List<StateEvent<T, C>> stateAllCallbacks = new ArrayList<>();
    private ContextJsonAdapter<C> contextJsonAdapter;
    private boolean enableEvents = true;

    public StaterStateMachine(List<Transition<T, C>> transitions, C context, T startState, Set<T> states, Map<String, List<TransitionMiddleware<C>>> transitionMiddlewares, List<TransitionNameMiddleware<C>> transitionAllMiddlewares, Map<String, List<Event<C>>> transitionCallbacks, List<NameEvent<C>> transitionAllCallbacks, Map<T, List<Event<C>>> stateCallbacks, List<StateEvent<T, C>> stateAllCallbacks, ContextJsonAdapter<C> contextJsonAdapter) {
        this.transitions = transitions;
        this.context = context;
        this.startState = startState;
        this.states = states;
        this.state = startState;
        this.transitionsGroupedStart = new HashMap<>();
        for (Transition<T, C> transition : transitions) {
            transitionsGroupedStart.computeIfAbsent(transition.start(), k -> new ArrayList<>()).add(transition);
        }
        this.transitionsByName = new HashMap<>();
        for (Transition<T, C> transition : transitions) {
            transitionsByName.put(transition.name(), transition);
        }
        this.transitionMiddlewares = transitionMiddlewares;
        this.transitionAllMiddlewares = transitionAllMiddlewares;
        this.transitionCallbacks = transitionCallbacks;
        this.transitionAllCallbacks = transitionAllCallbacks;
        this.stateCallbacks = stateCallbacks;
        this.stateAllCallbacks = stateAllCallbacks;
        this.contextJsonAdapter = contextJsonAdapter;
    }

    public StaterStateMachine(List<Transition<T, C>> transitions, C context, T startState) {
        this.transitions = transitions;
        this.context = context;
        this.startState = startState;
        this.states = Stream.concat(transitions.stream().map(Transition::start), transitions.stream().map(Transition::end)).collect(Collectors.toSet());
        this.state = startState;
        for (Transition<T, C> transition : transitions) {
            transitionsGroupedStart.computeIfAbsent(transition.start(), k -> new ArrayList<>()).add(transition);
        }
        for (Transition<T, C> transition : transitions) {
            transitionsByName.put(transition.name(), transition);
        }
    }

    public T getState() {
        return state;
    }

    public C getContext() {
        return context;
    }

    public void transition(String name) {
        Transition<T, C> transition = transitionsByName.get(name);
        if (transition == null) {
            throw new IllegalStateException("Transition not found: " + name);
        }
        if (!state.equals(transition.start())) {
            throw new IllegalStateException("Start state does not match transition's start state: " + transition.start());
        }

        Runnable conditionHandler = () -> {
            if (transition.condition() != null && !transition.condition().test(context)) {
                throw new IllegalStateException("Condition return false for transition " + name);
            }
        };

        List<TransitionMiddleware<C>> transitionMiddleware = transitionMiddlewares.get(name);
        Iterator<TransitionMiddleware<C>> middlewareIterator = (transitionMiddleware != null) ? transitionMiddleware.iterator() : Collections.emptyIterator();

        Consumer<C> internalNext = new Consumer<>() {
            @Override
            public void accept(C ctx) {
                if (middlewareIterator.hasNext()) {
                    middlewareIterator.next().apply(ctx, this::accept);
                } else {
                    conditionHandler.run();
                }
            }
        };

        Iterator<TransitionNameMiddleware<C>> allMiddlewareIterator = transitionAllMiddlewares.iterator();
        BiConsumer<String, C> next = new BiConsumer<>() {
            @Override
            public void accept(String transitionName, C ctx) {
                if (allMiddlewareIterator.hasNext()) {
                    allMiddlewareIterator.next().apply(transitionName, ctx, this::accept);
                } else {
                    internalNext.accept(ctx);
                }
            }
        };

        if (enableEvents) {
            next.accept(name, context);
        }

        state = transition.end();
        if (enableEvents && transition.event() != null) {
            transition.event().accept(context);
        }
        transitionCallbacks.getOrDefault(name, Collections.emptyList()).forEach(event -> event.accept(context));
        transitionAllCallbacks.forEach(event -> event.accept(name, context));
        stateCallbacks.getOrDefault(state, Collections.emptyList()).forEach(event -> event.accept(context));
        stateAllCallbacks.forEach(event -> event.accept(state, context));
    }

    public void autoTransition() {
        List<Transition<T, C>> possibleTransitions = transitionsGroupedStart.getOrDefault(state, Collections.emptyList());
        for (Transition<T, C> transition : possibleTransitions) {
            try {
                transition(transition.name());
                return;
            } catch (Exception ignored) {
            }
        }
    }

    public String toJsonSchema() throws Exception {
        Collator collator = Collator.getInstance(Locale.getDefault());
        collator.setStrength(Collator.PRIMARY);
        List<T> sortedStates = new ArrayList<>(states);
        sortedStates.sort(Comparator.comparing(T::toString, collator));
        ObjectMapper objectMapper = new ObjectMapper();
        return objectMapper.writeValueAsString(new JsonSchema<>(sortedStates, startState, transitions.stream().map(transition -> new Transition<T, EmptyContext>(transition.name(), transition.start(), transition.end(), null, null)).toList()));
    }

    public String toJson() throws Exception {
        if (contextJsonAdapter == null) {
            throw new IllegalStateException("ContextJsonAdapter is not set");
        }
        ObjectMapper objectMapper = new ObjectMapper();
        return objectMapper.writeValueAsString(new JsonState<>(state, contextJsonAdapter.toJson(context)));
    }

    public void fromJson(String json, Function<String, T> stateConverter) throws Exception {
        JsonState<T> jsonState = new ObjectMapper().readValue(json, new TypeReference<>() {
        });
        this.state = stateConverter.apply(jsonState.state().toString());
        this.context = contextJsonAdapter.fromJson(jsonState.context());
    }

    public void enableEvents() {
        this.enableEvents = true;
    }

    public void disableEvents() {
        this.enableEvents = false;
    }
}

File: projects/kotlin/src/main/kotlin/com/stater/statemachine/kotlin/StaterStateMachine.kt
package com.stater.statemachine.kotlin

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.text.Collator
import java.util.*


interface Context

class EmptyContext : Context

interface ContextJsonAdapter<C : Context> {
    fun toJson(context: C): String
    fun fromJson(json: String): C
}

typealias Event<C> = (context: C) -> Unit
typealias NameEvent<C> = (name: String, context: C) -> Unit
typealias StateEvent<T, C> = (state: T, context: C) -> Unit
typealias TransitionMiddleware<C> = (C, Event<C>) -> Unit
typealias TransitionNameMiddleware<C> = (String, C, NameEvent<C>) -> Unit
typealias StateMachineFactory<T, C> = (
    transitions: List<Transition<T, C>>,
    context: C,
    state: T,
    states: Set<T>,
    transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>>,
    transitionAllMiddlewares: List<TransitionNameMiddleware<C>>,
    transitionCallbacks: Map<String, List<Event<C>>>,
    transitionAllCallbacks: List<NameEvent<C>>,
    stateCallbacks: Map<T, List<Event<C>>>,
    stateAllCallbacks: List<StateEvent<T, C>>,
    contextJsonAdapter: ContextJsonAdapter<C>?,
) -> StaterStateMachine<T, C>

data class Transition<T, C : Context>(
    val name: String,
    val start: T,
    val end: T,
    @JsonIgnore val condition: ((C) -> Boolean)? = null,
    @JsonIgnore val event: Event<C>? = null,
)

data class JsonSchema<T>(
    val states: List<T>,
    val startState: T,
    val transitions: List<Transition<T, *>>,
)

data class JsonState<T>(
    val state: T,
    val context: String,
)

abstract class StaterStateMachine<T, C : Context>(
    private val transitions: List<Transition<T, C>>,
    private var context: C,
    private var startState: T = transitions[0].start,
    private val states: Set<T> = transitions.map { it.start }.toSet() + transitions.map { it.end }.toSet(),
    private var state: T = startState,
    private val transitionsGroupedStart: Map<T, List<Transition<T, C>>> = transitions.groupBy { it.start },
    private val transitionsByName: Map<String, Transition<T, C>> = transitions.associateBy { it.name },

    private val transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>> = mapOf(),
    private val transitionAllMiddlewares: List<TransitionNameMiddleware<C>> = listOf(),

    private val transitionCallbacks: Map<String, List<Event<C>>> = mapOf(),
    private val transitionAllCallbacks: List<NameEvent<C>> = listOf(),

    private val stateCallbacks: Map<T, List<Event<C>>> = mapOf(),
    private val stateAllCallbacks: List<StateEvent<T, C>> = listOf(),
    private var contextJsonAdapter: ContextJsonAdapter<C>? = null,
    private var enableEvents_: Boolean = true,
) {
    fun getState(): T = state
    fun getContext(): C = context

    fun transition(name: String) {
        val transition = transitionsByName[name] ?: error("Transition not found: $name")

        if (state != transition.start) {
            error("Start state does not match transition's start state: ${transition.start}")
        }

        fun conditionHandler() {
            transition.condition?.let {
                if (!it(context)) {
                    error("Condition return false for transition $name")
                }
            }
        }

        var index = 0
        var index2 = 0
        val transitionMiddleware = transitionMiddlewares[name]

        fun internalNext(context: C) {
            return if (transitionMiddleware != null && index2 < transitionMiddleware.size) {
                transitionMiddleware[index2++](context, ::internalNext)
            } else {
                conditionHandler()
            }
        }

        fun next(name: String, context: C) {
            return if (index < transitionAllMiddlewares.size) {
                transitionAllMiddlewares[index++](name, context, ::next)
            } else {
                internalNext(context)
            }
        }

        if (enableEvents_)
            next(name, context)

        state = transition.end
        if (!enableEvents_) return
        transition.event?.let { it(context) }

        transitionAllCallbacks.forEach { it(name, context) }
        transitionCallbacks[name]?.forEach { it(context) }
        stateAllCallbacks.forEach { it(state, context) }
        stateCallbacks[state]?.forEach { it(context) }
    }

    fun autoTransition() {
        (transitionsGroupedStart[state] ?: emptyList()).forEach {
            try {
                transition(it.name)
                return@forEach
            } catch (_: Exception) {
            }
        }
    }

    fun toJsonSchema(): String {
        val collator = Collator.getInstance(Locale.getDefault()).apply {
            strength = Collator.PRIMARY
        }
        return jacksonObjectMapper().writeValueAsString(
            JsonSchema(
                states = states.toList().sortedWith { a, b -> collator.compare(a.toString(), b.toString()) },
                startState = startState,
                transitions = transitions
            )
        ) ?: error("Cant serialize")
    }

    fun toJson() = jacksonObjectMapper().writeValueAsString(
        JsonState(
            state = state, context = contextJsonAdapter!!.toJson(context)
        )
    ) ?: error("Cant serialize")

    fun fromJson(json: String, stateConverter: (String) -> T) =
        jacksonObjectMapper().readValue<JsonState<T>>(json).let {
            this.state = stateConverter(it.state.toString())
            this.context = contextJsonAdapter!!.fromJson(it.context)
        }

    fun enableEvents() {
        enableEvents_ = true
    }

    fun disableEvents() {
        enableEvents_ = false
    }
}


class BaseFSM<T, C : Context>(
    transitions: List<Transition<T, C>>,
    context: C,
    startState: T,
    states: Set<T>,
    transitionMiddlewares: Map<String, List<TransitionMiddleware<C>>>,
    transitionAllMiddlewares: List<TransitionNameMiddleware<C>>,
    transitionCallbacks: Map<String, List<Event<C>>>,
    transitionAllCallbacks: List<NameEvent<C>>,
    stateCallbacks: Map<T, List<Event<C>>>,
    stateAllCallbacks: List<StateEvent<T, C>>,
    contextJsonAdapter: ContextJsonAdapter<C>?,
) : StaterStateMachine<T, C>(
    transitions = transitions,
    context = context,
    startState = startState,
    states = states,
    transitionMiddlewares = transitionMiddlewares,
    transitionAllMiddlewares = transitionAllMiddlewares,
    transitionCallbacks = transitionCallbacks,
    transitionAllCallbacks = transitionAllCallbacks,
    stateCallbacks = stateCallbacks,
    stateAllCallbacks = stateAllCallbacks,
    contextJsonAdapter = contextJsonAdapter
)


class StaterStateMachineBuilder<T, C : Context>(
    private val transitions: MutableMap<String, Transition<T, C>> = mutableMapOf(),
    private var state: T? = null,
    private var states: MutableSet<T> = mutableSetOf(),
    private var context: C? = null,

    private val transitionMiddlewares: MutableMap<String, MutableList<TransitionMiddleware<C>>> = mutableMapOf(),
    private val transitionAllMiddlewares: MutableList<TransitionNameMiddleware<C>> = mutableListOf(),

    private val transitionCallbacks: MutableMap<String, MutableList<Event<C>>> = mutableMapOf(),
    private val transitionAllCallbacks: MutableList<NameEvent<C>> = mutableListOf(),

    private val stateCallbacks: MutableMap<T, MutableList<Event<C>>> = mutableMapOf(),
    private val stateAllCallbacks: MutableList<StateEvent<T, C>> = mutableListOf(),
    private var factory: StateMachineFactory<T, C> = {
            transitionsA,
            contextA,
            startStateA,
            statesA,
            transitionMiddlewaresA,
            transitionAllMiddlewaresA,
            transitionCallbacksA,
            transitionAllCallbacksA,
            stateCallbacksA,
            stateAllCallbacksA,
            contextJsonAdapterA,
        ->
        BaseFSM(
            transitions = transitionsA,
            startState = startStateA,
            states = statesA,
            context = contextA,
            transitionMiddlewares = transitionMiddlewaresA,
            transitionAllMiddlewares = transitionAllMiddlewaresA,
            transitionCallbacks = transitionCallbacksA,
            transitionAllCallbacks = transitionAllCallbacksA,
            stateCallbacks = stateCallbacksA,
            stateAllCallbacks = stateAllCallbacksA,
            contextJsonAdapter = contextJsonAdapterA
        )
    },
    private var contextJsonAdapter: ContextJsonAdapter<C>? = null,
) {
    fun addTransition(
        name: String,
        start: T,
        end: T,
        condition: ((C) -> Boolean)? = null,
        event: ((C) -> Unit)? = null,
    ) = apply {
        addState(start)
        addState(end)
        transitions[name] = Transition(name, start, end, condition, event)
    }

    fun addState(state: T) = apply {
        if (!states.contains(state)) {
            states.add(state)
        }
    }

    fun setTransitionCondition(name: String, condition: ((C) -> Boolean)) = apply {
        val transition = transitions[name] ?: error("Transition not found: $name")
        transitions[name] = transition.copy(condition = condition)
    }

    fun setTransitionEvent(name: String, event: ((C) -> Unit)) = apply {
        val transition = transitions[name] ?: error("Transition not found: $name")
        transitions[name] = transition.copy(event = event)
    }

    fun transitionMiddleware(name: String, middleware: TransitionMiddleware<C>) = apply {
        val middlewares = transitionMiddlewares.getOrDefault(name, mutableListOf())
        middlewares.add(middleware)
        transitionMiddlewares[name] = middlewares
    }

    fun transitionAllMiddleware(middleware: TransitionNameMiddleware<C>) = apply {
        transitionAllMiddlewares.add(middleware)
    }

    fun subscribeOnTransition(name: String, callback: Event<C>) = apply {
        val callbacks = transitionCallbacks.getOrDefault(name, mutableListOf())
        callbacks.add(callback)
        transitionCallbacks[name] = callbacks
    }

    fun subscribeOnAllTransition(callback: NameEvent<C>) = apply { transitionAllCallbacks.add(callback) }

    fun subscribeOnState(state: T, callback: Event<C>) = apply {
        val callbacks = stateCallbacks.getOrDefault(state, mutableListOf())
        callbacks.add(callback)
        stateCallbacks[state] = callbacks
    }

    fun subscribeOnAllState(callback: StateEvent<T, C>) = apply { stateAllCallbacks.add(callback) }

    fun setStartState(state: T) = apply { this.state = state }
    fun setContext(context: C) = apply { this.context = context }

    fun fromJsonSchema(schema: String, stateConverter: (String) -> T) = apply {
        jacksonObjectMapper().readValue<JsonSchema<T>>(schema).let {
            it.states.forEach { it2 -> addState(stateConverter(it2.toString())) }
            it.transitions.forEach { it2 ->
                addTransition(
                    it2.name,
                    stateConverter(it2.start.toString()),
                    stateConverter(it2.end.toString())
                )
            }
            setStartState(stateConverter(it.startState.toString()))
        }
    }

    fun setFactory(factory: StateMachineFactory<T, C>) = apply { this.factory = factory }

    fun setContextJsonAdapter(contextJsonAdapter: ContextJsonAdapter<C>) = apply {
        this.contextJsonAdapter = contextJsonAdapter
    }

    fun build(): StaterStateMachine<T, C> {
        return factory(
            transitions.values.toList(),
            context ?: error("Context must be set"),
            state ?: transitions.values.toList()[0].start,
            states,
            transitionMiddlewares,
            transitionAllMiddlewares,
            transitionCallbacks,
            transitionAllCallbacks,
            stateCallbacks,
            stateAllCallbacks,
            contextJsonAdapter
        )
    }
}